usr/man/man/usr/share/man/man5/charmap.5
'\" te
.\"  Copyright (c) 1992, X/Open Company Limited  All Rights Reserved  Portions Copyright (c) 2003, Sun Microsystems, Inc.  All Rights Reserved
.\"  Sun Microsystems, Inc. gratefully acknowledges The Open Group for permission to reproduce portions of its copyrighted documentation. Original documentation from The Open Group can be obtained online at  http://www.opengroup.org/bookstore/.
.\" The Institute of Electrical and Electronics Engineers and The Open Group, have given us permission to reprint portions of their documentation. In the following statement, the phrase "this text" refers to portions of the system documentation. Portions of this text are reprinted and reproduced in electronic form in the Sun OS Reference Manual, from IEEE Std 1003.1, 2004 Edition, Standard for Information Technology -- Portable Operating System Interface (POSIX), The Open Group Base Specifications Issue 6, Copyright (C) 2001-2004 by the Institute of Electrical and Electronics Engineers, Inc and The Open Group. In the event of any discrepancy between these versions and the original IEEE and The Open Group Standard, the original IEEE and The Open Group Standard is the referee document. The original Standard can be obtained online at http://www.opengroup.org/unix/online.html.  This notice shall appear on any product containing this material. 
.TH charmap 5 "1 Dec 2003" "SunOS 5.11" "Standards, Environments, and Macros"
.SH NAME
charmap \- character set description file
.SH DESCRIPTION
.sp
.LP
A character set description file or \fIcharmap\fR defines characteristics for a coded character set. Other information about the coded character set may also be in the file. Coded character set character values are defined using symbolic character names followed by character encoding values.
.sp
.LP
The character set description file provides:
.RS +4
.TP
.ie t \(bu
.el o
The capability to describe character set attributes (such as collation order or character classes) independent of character set encoding, and using only the characters in the portable character set. This makes it possible to create generic \fBlocaledef\fR(1) source files for all codesets that share the portable character set.
.RE
.RS +4
.TP
.ie t \(bu
.el o
Standardized symbolic names for all characters in the portable character set, making it possible to refer to any such character regardless of encoding.
.RE
.SS "Symbolic Names"
.sp
.LP
Each symbolic name  is included in the file and is mapped to a unique encoding value (except for those symbolic names that are shown with identical glyphs). If the control characters commonly associated with the symbolic names in the following table are supported by the implementation, the symbolic names and their corresponding encoding values are included in the file. Some of the encodings associated with the symbolic names in this table may be the same as characters in the portable character set table.
.sp

.sp
.TS
tab() box;
lw(.92i) lw(.92i) lw(.92i) lw(.92i) lw(.92i) lw(.92i) 
lw(.92i) lw(.92i) lw(.92i) lw(.92i) lw(.92i) lw(.92i) 
.
<ACK><DC2><ENQ><FS><IS4><SOH>
<BEL><DC3><EOT><GS><LF><STX>
<BS><DC4><ESC><HT><NAK><SUB>
<CAN><DEL><ETB><IS1><RS><SYN>
<CR><DLE><ETX><IS2><SI><US>
<DC1><EM><FF><IS3><SO><VT>
.TE

.SS "Declarations"
.sp
.LP
The following declarations can precede the character definitions. Each must consist of the symbol shown in the following list, starting in column 1, including the surrounding brackets, followed by one or more blank characters, followed by the value to be assigned to the symbol.
.sp
.ne 2
.mk
.na
\fB<\fIcode_set_name\fR>\fR
.ad
.RS 19n
.rt  
The name of the coded character set for which the character set description file is defined.
.RE

.sp
.ne 2
.mk
.na
\fB<\fImb_cur_max\fR>\fR
.ad
.RS 19n
.rt  
The maximum number of bytes in a multi-byte character. This defaults to \fB1\fR.
.RE

.sp
.ne 2
.mk
.na
\fB<\fImb_cur_min\fR>\fR
.ad
.RS 19n
.rt  
An unsigned positive integer value that defines the minimum number of bytes in a character for the encoded character set.
.RE

.sp
.ne 2
.mk
.na
\fB<\fIescape_char\fR>\fR
.ad
.RS 19n
.rt  
The escape character used to indicate that the characters following will be interpreted in a special way, as defined later in this section. This defaults to backslash ('\fB\e\fR\&'), which is the character glyph used in all the following text and examples, unless otherwise noted.
.RE

.sp
.ne 2
.mk
.na
\fB<\fIcomment_char\fR>\fR
.ad
.RS 19n
.rt  
The character that when placed in column 1 of a charmap line, is used to indicate that the line is to be ignored. The default character is the number sign (#).
.RE

.SS "Format"
.sp
.LP
The character set mapping definitions will be all the lines immediately following an identifier line containing the string \fBCHARMAP\fR starting in column 1, and preceding a trailer line containing the string \fBEND\fR \fBCHARMAP\fR starting in column 1. Empty lines and lines containing a \fI<comment_char>\fR in the first column will be ignored. Each non-comment line of the character set mapping definition, that is, between the \fBCHARMAP\fR and \fBEND CHARMAP\fR lines of the file), must be in either of two forms:
.sp
.in +2
.nf
\fB"%s %s %s\en",\fR<\fIsymbolic-name\fR>,<\fIencoding\fR>,<\fIcomments\fR>
.fi
.in -2

.sp
.LP
or
.sp
.in +2
.nf
\fB"%s...%s %s %s\en",\fR<\fIsymbolic-name\fR>,<\fIsymbolic-name\fR>, <\fIencoding\fR>,\e
         <\fIcomments\fR>
.fi
.in -2

.sp
.LP
In the first format, the line in the character set mapping definition defines a single symbolic name and a corresponding encoding. A character following an escape character is interpreted as itself; for example, the sequence "<\fB\e\e\e\fR>>" represents the symbolic name "\fI\e>\fR" enclosed between angle brackets.
.sp
.LP
In the second format, the line in the character set mapping definition defines a range of one or more symbolic names. In this form, the symbolic names must consist of zero or more non-numeric characters,  followed by an integer formed by one or more decimal digits. The characters preceding the integer must be identical in the two symbolic names, and the integer formed by the digits in the second symbolic name must be equal to or greater than the integer formed by the digits in the first name. This is interpreted as a series of symbolic names formed from the common part and each of the integers between the first and the second integer, inclusive. As an example, \fB<j0101>...<j0104>\fR is interpreted as the symbolic names \fB<j0101>\fR, \fB<j0102>\fR, \fB<j0103>\fR, and \fB<j0104>\fR, in that order.
.sp
.LP
A character set mapping definition line must exist for all symbolic names and must define the coded character value that corresponds to the character glyph indicated in the table, or the coded character value that corresponds with the control character symbolic name. If the control characters commonly associated with the symbolic names  are supported by the implementation, the symbolic name and the corresponding encoding value must be included in the file. Additional unique symbolic names may be included. A coded character value can be represented by more than one symbolic name.
.sp
.LP
The encoding part is expressed as one (for single-byte character values) or more concatenated decimal, octal or hexadecimal constants in the following formats:
.sp
.in +2
.nf
\fB"%cd%d",\fR<\fIescape_char\fR>,<\fIdecimal byte value\fR>

\fB"%cx%x",\fR<\fIescape_char\fR>,<\fIhexadecimal byte value\fR>

\fB"%c%o",\fR<\fIescape_char\fR>,<\fIoctal byte value\fR>
.fi
.in -2

.SS "Decimal Constants"
.sp
.LP
Decimal constants must be represented by two or three decimal digits, preceded by the escape character and the lower-case letter \fBd\fR; for example, \fB\ed05\fR, \fB\ed97\fR, or \fB\ed143\fR\&. Hexadecimal constants must be represented by two hexadecimal digits, preceded by the escape character and the lower-case letter \fBx\fR; for example, \fB\ex05\fR, \fB\ex61\fR, or \fB\ex8f\fR\&. Octal constants must be represented by two or three octal digits, preceded by the escape character; for example, \fB\e05\fR, \fB\e141\fR, or \fB\e217\fR\&. In a portable charmap file, each constant must represent an 8-bit byte. Implementations supporting other byte sizes may allow constants to represent values larger than those that can be represented in 8-bit bytes, and to allow additional digits in constants. When constants are concatenated for multi-byte character values, they must be of the same type, and interpreted in byte order from first to last with the least significant byte of the multi-byte character specified by the last constant.
.SS "Ranges of Symbolic Names"
.sp
.LP
In lines defining ranges of symbolic names, the encoded value is the value for the first symbolic name in the range (the symbolic name preceding the ellipsis). Subsequent symbolic names defined by the range will have encoding values in increasing order. Bytes are treated as unsigned octets and carry is propagated between the bytes as necessary to represent the range. However, because this causes a null byte in the second or subsequent bytes of a character, such a declaration should not be specified. For example, the line
.sp
.in +2
.nf
<j0101>...<j0104>     \ed129\ed254
.fi
.in -2

.sp
.LP
is interpreted as:
.sp
.in +2
.nf
<j0101>                \ed129\ed254
<j0102>                \ed129\ed255
<j0103>                \ed130\ed00
<j0104>                \ed130\ed01
.fi
.in -2

.sp
.LP
The expanded declaration of the symbol \fB<j0103>\fR in the above example is an invalid specification, because it contains a null byte in the second byte of a character.
.sp
.LP
The comment is optional.
.SS "Width Specification"
.sp
.LP
The following declarations can follow the character set mapping definitions (after the "\fBEND CHARMAP\fR" statement). Each consists of the keyword shown in the following list, starting in column 1, followed by the value(s) to be associated to the keyword, as defined below.
.sp
.ne 2
.mk
.na
\fB\fBWIDTH\fR\fR
.ad
.RS 17n
.rt  
A non-negative integer value defining the column width for the printable character in the coded character set mapping definitions. Coded character set character values are defined using symbolic character names followed by column width values. Defining a character with more than one \fBWIDTH\fR produces undefined results. The \fBEND WIDTH\fR keyword is used to terminate the \fBWIDTH\fR definitions. Specifying the width of a non-printable character in a \fBWIDTH\fR declaration produces undefined results.
.RE

.sp
.ne 2
.mk
.na
\fB\fBWIDTH_DEFAULT\fR\fR
.ad
.RS 17n
.rt  
A non-negative integer value defining the default column width for any printable character not listed by one of the \fBWIDTH\fR keywords. If no \fBWIDTH_DEFAULT\fR keyword is included in the charmap, the default character width is \fB1\fR.
.RE

.sp
.LP
Example:
.sp
.LP
After the "\fBEND CHARMAP\fR" statement, a syntax for a width definition would be:
.sp
.in +2
.nf
WIDTH
<A>             1
<B>             1
<C>...<Z>       1
\&...
<fool>...<foon> 2
\&...
END WIDTH
.fi
.in -2
.sp

.sp
.LP
In this example, the numerical code point values represented by the symbols \fB<A>\fR and \fB<B>\fR are assigned a width of \fB1\fR. The code point values \fB< C>\fR to \fB<Z>\fR inclusive, that is, \fB<C>\fR, \fB<D>\fR, \fB<E>\fR, and so on, are also assigned a width of \fB1\fR. Using \fB<A>. . .<Z>\fR would have required fewer lines, but the alternative was shown to demonstrate flexibility. The keyword \fBWIDTH_DEFAULT\fR could have been added as appropriate.
.SH SEE ALSO
.sp
.LP
\fBlocale\fR(1), \fBlocaledef\fR(1), \fBnl_langinfo\fR(3C), \fBextensions\fR(5), \fBlocale\fR(5) 


usr/closed/cmd/pax/charmap.c
/*
 * COPYRIGHT NOTICE
 *
 * This source code is designated as Restricted Confidential Information
 * and is subject to special restrictions in a confidential disclosure
 * agreement between HP, IBM, SUN, NOVELL and OSF.  Do not distribute
 * this source code outside your company without OSF's specific written
 * approval.  This source code, and all copies and derivative works
 * thereof, must be returned or destroyed at request. You must retain
 * this notice on any copies which you make.
 *
 * (c) Copyright 1990, 1991, 1992, 1993 OPEN SOFTWARE FOUNDATION, INC.
 * ALL RIGHTS RESERVED
 */
/*
 * OSF/1 1.2
 */
#if !defined(lint) && !defined(_NOIDENT)
	static char rcsid[] = "@(#)$RCSfile: charmap.c,v $ $Revision: 1.3.1.2 "
	    "$ (OSF) $Date: 1991/10/09 14:12:00 $";
#endif
/*
 * charmap.c - functions to parse and store charmap files.
 *
 * DESCRIPTION
 *
 *	Routines to read and store a charmap file as specified
 *	in POSIX 1003.2 Draft 11.  Also contains the routines
 * 	to access the charmap and convert filenames.
 *
 * AUTHOR
 *
 *     Tom Jordahl - The Open Software Foundation
 *
 *
 */

/* Headers */


/*
 * File wide globals
 */

static char	 line[LMAX];		/* buffer to hold each line */
static char	*lineptr;		/* location in line buffer */
char		escape_char = '\\'; 	/* escape charater (default '\') */
int		mbmaxlen = 1;		/* max bytes in multi-byte */
int		mbminlen = 1;		/* min bytes in multi-byte */
Charentry	*char_head = NULL;	/* head of Charentry list */
Charentry	*char_tail = NULL;	/* tail of Charentry list */
int		entry_count = 0;	/* number of entries in charmap */
Charentry	*charmap;		/* The final sorted charmap table */
Charentry	search_key;		/* Seach key for bsearch */



static char	*get_name(void);
static int	make_entry(char *, Value *);
static int	do_range(char *, char *, Value *);
static int	get_encoding(Value *);
static int	lookup(char *);
static void	charmap_error(char *, int);
static void	eatspace(void);
static void	sort_table(void);
static char	*mblookup(char *);
static char 	*symlookup(char *, int *);
int		sym_compare(const Charentry *, const Charentry *);
int		bval_compare(const Charentry *, const Charentry *);
int 		byte_match(const Charentry *k1, const Charentry *k2);






/*
 * read_charmap - main routine which reads a charmap file
 *
 * DESCRIPTION
 *
 *	Parses the charmap file, storing everything we need to
 *	support the -e option of pax.
 *
 * PARAMETERS
 *
 *	char 	*file 	- Name of the charmap file.
 *
 * RETURNS
 *
 * 	0 	- if all went well
 * 	-1	- the charmap file was (in our eyes) bogus.
 *
 */


int
read_charmap(char *file)
{
	FILE	*cfile;			/* charmap input file */
	int	 linenum = 0;		/* line number in the charmap file */
	char	*name1;			/* first symbolic name in line */
	char	*name2;			/* second sybbolic name (if range) */
	char	 comment_char = '#'; 	/* comment charater (default '#') */
	short	 range = FALSE;		/* true if we are parsing a range */
	short	 incharmap = FALSE;	/* true if we have parsed CHARMAP */
	Value	 value;			/* struct that hold both mb and wc */
	char	*charmap = "CHARMAP";
	char	*endcharmap = "END CHARMAP";
	int		 charmap_len = strlen(charmap);
	int		 endcharmap_len = strlen(endcharmap);


	if ((cfile = fopen(file, "r")) == NULL) {
		warn(file, strerror(errno));
		return (-1);
	}

	while (fgets(line, LMAX, cfile) != NULL) {
		linenum++;
		lineptr = line;
		range = FALSE;

		if ((line[0] == comment_char) || (line[0] == '\n'))
			continue;		/* comment */

		if (strncmp(line, charmap, charmap_len) == 0) {
			incharmap = TRUE;
			continue;
		}

		if (strncmp(line, endcharmap, endcharmap_len) == 0) {
			incharmap = FALSE;
			break;	/* we are done with the file */
		}

		if ((line[0] != '<') || ((name1 = get_name()) == NULL)) {
			charmap_error(MSGSTR(CM_SYNTAX,
			    "Syntax error - skipping"), linenum);
		    continue;
		}

		if (!incharmap) {
			eatspace();
			switch (lookup(name1)) {

			case CODE_SET_NAME:
				break;

			case MB_MAX:
				mbmaxlen = atoi(lineptr);
				break;

			case MB_MIN:
				mbminlen = atoi(lineptr);
				break;

			case ESCAPE_CHAR:
				escape_char = *lineptr;
				break;

			case COMMENT_CHAR:
				comment_char = *lineptr;
				break;

			default:
				charmap_error(MSGSTR(CM_DECL,
				    "Error in declarations - skipping"),
				    linenum);
				free(name1);
				break;
			}
			continue;
		}

		if (*lineptr == '.' && *++lineptr == '.' && *++lineptr == '.') {
			lineptr++;
			range = TRUE;
			if ((name2 = get_name()) == NULL) {
				charmap_error(MSGSTR(CM_RANGE,
				    "Error in range - skipping"), linenum);
				free(name1);
				continue;
			}
		}

		if (get_encoding(&value) < 0) {
			charmap_error(MSGSTR(CM_ENCODE,
			    "Error in encoding - skipping"), linenum);
			free(name1);
			continue;
		}

		if (range) {
			if (do_range(name1, name2, &value) < 0) {
				charmap_error(MSGSTR(CM_ADDRANGE,
				    "Error adding range"), linenum);
				free(name1);
				free(name2);
				continue;
			}
			free(name2);
		} else if (make_entry(name1, &value) < 0) {
			charmap_error(MSGSTR(CM_ADDENTRY,
			    "Error adding entry"), linenum);
			free(name1);
			continue;
		}
		free(name1);
	}
	sort_table();
	return (0);
}


/*
 * charmap_convert - run a filename through the charmap conversion.
 *
 * DESCRIPTION
 *
 * 	Converts a filename to or from the POSIX portable charater set
 *
 *  -w  Examines the filename string for any characters which are
 * 	not in the POSIX portable character set.  If we find one
 * 	we look it up in our charmap table and replace it with
 * 	its symbol name.  ie '/' would be replaced with "<slash>".
 *
 * [-r] Examines the filename for any symbol names and replaces
 *	them with their encoding from the charmap.  If we can't
 *	find the symbol, we will just leave it alone.
 *
 *
 * PARAMETERS
 *
 *	char	*name	- File name to scan.
 *			  Must be a pointer to at least PATH_MAX + 1 chars.
 *
 * RETURNS
 *
 *	0  - if succesfully converted the filename using the charmap.
 * 	-1 - if failure.
 *
 */


int
charmap_convert(char *name)
{
	char	buf[PATH_MAX+1];	/* scratch buffer */
	char	*bindex;		/* index in to buf */
	char	*nindex;		/* index in to name */
	char	c;			/* scratch character */
	char	*sym;			/* symbol returned from mblookup */
	char	*symbol;		/* symbol scrounged out of name */
	char	*symptr;		/* index in to symbol */
	short	esc;			/* flag for escape char */
	char	*orig_name;		/* original name for error */
	char	*val;			/* multi-byte value from symlookup */
	int	bytes;			/* numbers of bytes in 'val' */


	orig_name = mem_str(name);
	bindex = buf;
	nindex = name;
	if (f_create) {			/* -w */
		/*
		 *  insert escape chars before <,>, and escape_chars
		 */
		for (nindex = name; *nindex != '\0'; nindex++) {
			c = *nindex;
			if ((c == '<') || (c == '>') || (c == escape_char))
				*bindex++ = escape_char;
			*bindex++ = c;
			if ((bindex - buf) > PATH_MAX+1) {
				warn(orig_name, MSGSTR(CM_LONG,
				    "Filename becomes too long"));
				free(orig_name);
				return (-1);
			}
		}
		*bindex = '\0';

		/*
		 * filename now in buf
		 *
		 * Scan filename for any character not in PCS and replace with
		 * name.  The name we get back from mblookup will have <, >, and
		 * escape_chars escaped.
		 */
		nindex = name;		/* name is now our "scratch" buffer */
		for (bindex = buf; *bindex != '\0'; bindex++) {
			if (!isportable(*bindex)) {
				sym = mblookup(bindex);
				bindex += (mblen(bindex, MB_LEN_MAX) - 1);
				if (sym == NULL) {
					warn(orig_name, MSGSTR(CM_NOTFOUND,
					    "Could not find charmap entry"));
					free(orig_name);
					return (-1);
				}
				if ((nindex-name) + strlen(sym) > PATH_MAX+1) {
					warn(orig_name, MSGSTR(CM_LONG,
					    "Filename becomes too long"));
					free(orig_name);
					return (-1);
				}
				strcpy(nindex, sym);
				nindex += strlen(sym);
			} else
				*nindex++ = *bindex;
		}
		*nindex = '\0';
	} else {			/* -r or index */
		/* symbols can't be longer than the name itself */
		symbol = mem_get(strlen(name)+1);

		/* scan for a string "<string>" in name */
		for (nindex = name; *nindex != '\0'; nindex++) {
			c = *nindex;
			if (c == escape_char) {
				*bindex++ = c;
				esc = (esc + 1) % 2;
			} else if (c == '<' && !esc) {
				/* start of a new symbol */
				nindex++;			/* skip '<' */
				for (symptr = symbol; *nindex != '>' || esc;
				    nindex++) {
					c = *nindex;
					if (c == escape_char) {
						if (esc) {
							*symptr++ = escape_char;
							esc = 0;
						} else
							esc++;
					} else {
						if (esc && (c != '<') &&
						    (c != '>')) {
							*symptr++ = escape_char;
							esc = 0;
						}
						*symptr++ = c;
					}
				}
				symptr = '\0';

				/*
				 * lookup the symbol in charmap and replace
				 * with encoding
				 */
				val = symlookup(symbol, &bytes);
				if (val) {
					/* insert mb sequence */
					strncpy(bindex, val, bytes);
					bindex += bytes;
				} else {
					/* put back the symbol */
					strcpy(bindex, symbol);
					bindex += strlen(symbol);
				}
			} else {
				esc = 0;
				*bindex++ = c;
			}
		}
		*bindex = '\0';

		/*
		 * filename is now in buf - copy it back to name
		 *
		 * remove escape chars in front of <, >, and escape_chars
		 */
		for (esc = 0, bindex = buf, nindex = name; *bindex != '\0';
		    bindex++) {
			c = *bindex;
			if (c == escape_char) {
				if (esc) {
					*nindex++ = escape_char;
					esc = 0;
				} else
					esc++;
			} else {
				if (esc && (c != '<') && (c != '>'))
					*nindex++ = escape_char;
				esc = 0;
				*nindex++ = c;
			}
		}
		if (esc)
			*nindex++ = escape_char;
		*nindex = '\0';
		free(symbol);
	}
	free(orig_name);
	return (0);
}

/*
 * get_name - read a symbolic name of a character or declaration
 *
 * DESCRIPTION
 *
 *	Examines the line buffer (pointed at by lineptr) and returns
 * 	the string in between the first '<' and '>'.
 *
 *
 * PARAMETERS
 *
 * RETURNS
 *
 *	pointer to a symbolic name or NULL if name isn't found.
 *
 */


static char *
get_name(void)
{
	char	*ret;		/* pointer to returned buffer */
	char	*end;		/* pointer to the end of symbol */
	int	length = 0;	/* length of symbol */
	int	escaped = 0;	/* flag for escaped chars */
	int	i;		/* loop counter */

	while ((*lineptr != '<') && (*lineptr != '\0')) {
		lineptr++;	/* find bracket */
	}

	if (*lineptr == '\0')
		return (NULL);

	lineptr++;

	end = lineptr;
	while (*end && *end != '>') {
		length++;
		if (*end == escape_char)
			end += 2;
		else
			end++;
	}
	if (*end == '\0')
		return (NULL);

	if ((ret = mem_get(length+1)) == NULL)
		fatal(MSGSTR(NOMEM, "Out of memory"));

	for (i = 0; lineptr < end; lineptr++) {
		if ((*lineptr == escape_char) && !escaped) {
			escaped++;
			continue;
		} else {
			if (escaped)
				escaped = 0;
			ret[i++] = *lineptr;
		}
	}
	ret[length] = '\0';
	lineptr = end + 1;		/* just after '>' */
	return (ret);
}



/*
 * make_entry - Make an entry in the charmap table.
 *
 * DESCRIPTION
 *
 *	Makes an entry in the charmap for the given symbol name
 * 	with the given value.
 *
 *
 * PARAMETERS
 *
 *	char	 *symbol 	- symbol to make the entry for.
 *	Value	 *value 	- encoding value of this symbol.
 *
 * RETURNS
 *
 *	0	- OK
 *	-1 	- if error.
 *
 */


static int
make_entry(char *symbol, Value *value)
{
	Charentry	*entry;

	if ((entry = (Charentry *)mem_get(sizeof (Charentry))) == NULL)
		return (-1);

	entry->symbol = mem_str(symbol);
	strcpy((char *)&entry->value, (char *)&value->mbvalue);
	entry->nbytes = value->nbytes;
	entry->next = NULL;

	if (char_head == NULL) {
		char_head = entry;
		char_tail = entry;
	} else {
		char_tail->next = entry;
		char_tail = entry;
	}

	entry_count++;
	return (0);
}



/*
 * do_range - make the entries for a range of symbolic names.
 *
 * DESCRIPTION
 *
 *
 * 	Makes entries into the charmap for a range of characters
 * 	which have sequential values.
 *
 *	The symbol must names consist of zero or more nonnumeric characters
 *	followed by an interget formed by one or more decimal digits.
 *
 * PARAMETERS
 *
 *	char *name1 	- fist symbolic name
 *	char *name2 	- last symbolic name
 *	Value *value 	- starting value.
 *
 * RETURNS
 *
 *	0 	- is successfull.
 *	-1 	- if an error occurred.
 *
 */


static int
do_range(char *name1, char *name2, Value *value)
{
	int	start;		/* starting symbol number */
	int	end;		/* ending symbol number */
	int	numlen1;	/* length of number in name1 */
	int	numlen2;	/* length of number in name2 */
	char	*buf;		/* area of memory for constructed name */
	char	*prefix;	/* nondigit part of name */
	char	*p;
	int	i;
	int	byte;


	p = name1;
	while (*p && !isdigit(*p))
		p++;
	if (!*p)
		return (-1);
	start = atoi(p);
	numlen1 = strlen(p);

	p = name2;
	while (*p && !isdigit(*p))
		p++;
	if (!*p)
		return (-1);
	end = atoi(p);
	numlen2 = strlen(p);

	*p = '\0';
	prefix = name2;
	if ((start > end) || (numlen1 != numlen2))
		return (-1);

	buf = mem_get(strlen(name1));

	for (i = start; i <= end; i++) {
		sprintf(buf, "%s%0*d", prefix, numlen1, i);

		if (make_entry(buf, value) < 0)
			return (-1);

		byte = value->nbytes - 1;
		value->mbvalue[byte]++;
		for (; byte > 0; byte--) {
			if ((value->mbvalue[byte]) == 0)
				value->mbvalue[byte-1]++;
		}
	}
	free(buf);
	return (0);
}


/*
 * get_encoding - Decodes the character encoding into multi-bye.
 *
 * DESCRIPTION
 *
 *	Examines the line buffer (via lineptr)  and returns the
 * 	multi-byte encoding of the one or
 * 	more concatinated decimal, hex, or octal numbers.
 *
 *
 * PARAMETERS
 *
 *	Value	*value	- where to put the encoding.
 *
 * RETURNS
 *
 *	0 	- OK
 * 	-1 	- on error.
 *
 */


static int
get_encoding(Value *value)
{
	char	c;
	int	i;
	int	code;

	eatspace();

	for (i = 0; i < mbmaxlen; i++) {
		if ((*lineptr == '\0') || (*lineptr++ != escape_char))
			break;

		switch ((c = *lineptr++)) {

		case 'd':			/* decimal */
			code = atoi(lineptr);
			break;

		case 'x':			/* hex */
			code = strtol(lineptr, (char **)NULL, 16);
			break;

		default:
			if (isdigit(c) && (c >= '0' && c <= '7')) {
				/* octal */
				code = strtol(lineptr, (char **)NULL, 8);
			    } else
				return (-1);
		}
		value->mbvalue[i] = (char)code;
	}
	if (i < mbminlen)
		return (-1);

	value->nbytes = i;

	return (0);
}


/*
 * lookup - lookup a symbolic name for a declaration
 *
 * DESCRIPTION
 *
 *	Takes a string and maches it with one of the declaration
 * 	names.
 *
 *
 * PARAMETERS
 *
 *	char *name 	- name to look at.
 *
 * RETURNS
 *
 *	Returns an integer ordinal value of the name
 *	-1 if name isn't found.
 *
 */


static int
lookup(char *name)
{

	if (strcmp(name, "code_set_name") == 0)
		return (CODE_SET_NAME);
	if (strcmp(name, "mb_cur_max") == 0)
		return (MB_MAX);
	if (strcmp(name, "mb_cur_min") == 0)
		return (MB_MIN);
	if (strcmp(name, "escape_char") == 0)
		return (ESCAPE_CHAR);
	if (strcmp(name, "comment_char") == 0)
		return (COMMENT_CHAR);

	return (-1);
}


/*
 * charmap_error - print an error message
 *
 * DESCRIPTION
 *
 *	Prints the line number and a provided message.
 *
 *
 * PARAMETERS
 *
 *	char *mess 	- message to print
 *	int   line	- line number where error occurred.
 *
 */


static void
charmap_error(char *mess, int l)
{
	char	lbuf[10];

	sprintf(lbuf, MSGSTR(CM_LINE, "line %d"), l);
	warn(lbuf, mess);
}

/*
 * eatspace - move lineptr past spaces
 *
 */

static void
eatspace(void)
{
	while ((*lineptr != '\0') && isspace(*lineptr))
		lineptr++;
}


/*
 * sort_table - sort the character map.
 *
 * DESCRIPTION
 *
 *	Reads through the linked list we created while parsing
 * 	the charmap file and constructs a contiguous array
 * 	of the entryies, which is sorts.
 *
 * 	Two kinds of sorts are done:
 *	 - when writing an archive, we sort on the multi-byte encoding
 *	 - when reading an archive, we sort on the symbol name.
 *
 *
 */


static void
sort_table(void)

{
	Charentry	*cptr;		/* list walker */
	int		i = 0;


	charmap = (Charentry *)mem_get(entry_count * sizeof (Charentry));
	if (!charmap)
		fatal(MSGSTR(CM_NOMEM, "Out of memory constructing charmap"));

	cptr = char_head;
	while (cptr != NULL) {
		memcpy(&charmap[i], cptr, sizeof (Charentry));
		free(cptr);
		cptr = charmap[i].next;
		i++;
	}

	/* sort them */
	if (f_create)
		/* sort by byte value */
		qsort(charmap, (size_t)entry_count, sizeof (Charentry),
		    (int (*)(const void *, const void *))bval_compare);
	else		 /* extract or list */
		/* sort by symbol name */
		qsort(charmap, (size_t)entry_count, sizeof (Charentry),
		    (int (*)(const void *, const void *))sym_compare);
}


/*
 * mblookup - lookup a symbolic name of a multi-byte character.
 *
 * DESCRIPTION
 *
 *	Searches through the charmap for a multi-byte character
 * 	that matches the one pointed to by mbchar.
 *
 *	before returning the symbol name we place an escape in
 * 	front of '<', '>', or escape_char's.  We also put in
 * 	the surrounding brackets '<>'.
 *
 * PARAMETERS
 *
 *	char	*mbchar	- pointer to the start of a multi-byte char.
 *
 * RETURNS
 *
 *	pointer to a symbolic name or NULL if name isn't found.
 *
 */


static char *
mblookup(char *mbchar)
{
	Charentry	*entry;			/* entry return by bsearch */
	char		*symbol;		/* symbol from entry */
	char		escname[PATH_MAX + 1];	/* buffr to hold final symbol */
	char		*eindex;		/* index into escname buffer */
	char		*ret;			/* pointer to returned buffer */
	char		c;			/* scratch character */

	strncpy((char *)search_key.value, mbchar, 10);
	entry = bsearch(&search_key, charmap, (size_t)entry_count,
	    sizeof (Charentry),
	    (int (*)(const void *, const void *))byte_match);
	if (entry == NULL)
		return (NULL);

	symbol = entry->symbol;
	eindex = escname;
	*eindex++ = '<';
	for (; *symbol != '\0'; symbol++) {
		c = *symbol;
		if ((c == '<') || (c == '>') || (c == escape_char))
			*eindex++ = escape_char;
		*eindex++ = c;
	}
	*eindex++ = '>';
	*eindex = '\0';

	ret = mem_str(escname);
	return (ret);
}


/*
 * symlookup - lookup the multi-byte encoding of a symbolic name
 *
 * DESCRIPTION
 *
 *	Searches through the charmap for a matching symbol
 * 	that matches the one pointed to by mbchar.  Returns
 * 	that multi-byte encoding of that symbol from the charmap.
 *
 * PARAMETERS
 *
 *	char	*symbol	- symbol name to search charmap for
 * 	int	*nbytes - number of valid bytes in the returned mb value.
 *
 * RETURNS
 *
 *	pointer to a value array of bytes  or NULL if symbol isn't found.
 *
 */


static char *
symlookup(char *symbol, int *nbytes)
{
	Charentry	*entry;

	search_key.symbol = symbol;
	entry = bsearch(&search_key, charmap, (size_t)entry_count,
		    sizeof (Charentry),
		    (int (*)(const void *, const void *))sym_compare);
	if (entry == NULL)
		return (NULL);

	*nbytes = entry->nbytes;
	return ((char *)entry->value);
}


/*
 * byte_match - comparison routine for bsearch to find a matching mbvalue
 *
 * DESCRIPTION
 *
 * 	We need to know which of the two elements is the key since
 * 	we must use the nbytes from the element in the table.
 */

int
byte_match(const Charentry *k1, const Charentry *k2)
{
	const Charentry	*item;

	if (k1 == &search_key)
		item = k2;
	else
		item = k1;

	return (strncmp((char *)k1->value, (char *)k2->value, item->nbytes));
}

/*
 * bval_compare - comparison routine for qsort-ing multi-byte encodings
 *
 * DESCRIPTION
 *
 * 	Used for qsort to compare two multi-byte encodings
 *
 * RETURNS
 *	-1 - if x1 < x2
 *	 0 - if x1 = x2
 *	 1 - if x1 > x2
 *
 */
int
bval_compare(const Charentry *x1, const Charentry *x2)
{

	if (x1->nbytes > x2->nbytes)
		return (1);
	else if (x1->nbytes < x2->nbytes)
		return (-1);
	else
		return (strncmp((char *)x1->value,
		    (char *)x2->value, x1->nbytes));
}

/*
 * sym_compare - comparison routine for symbols
 *
 * DESCRIPTION
 *
 * 	Used for qsort and bsearch to compare two symbols
 *
 * RETURNS
 *	-1 - if x1 < x2
 *	 0 - if x1 = x2
 *	 1 - if x1 > x2
 *
 */
int
sym_compare(const Charentry *x1, const Charentry *x2)
{

	return (strcmp(x1->symbol, x2->symbol));
}
